### DISCUSSION TITLE
Are the Score Director methods for adding/removing entities truly deprecated?

### DISCUSSION THREAD

**Discussion by storm-TOS** [TimefoldAI/timefold-solver#1154] [General]
Certain Score Director methods for managing a variable number of planning entities (before/afterEntityAdded/Removed) have been marked as deprecated as of Timefold version 1.8. However, about half a year ago, @ge0ffrey suggested (in private communication) that this feature might receive continued support after all. What are your current plans regarding this deprecation?

Our task scheduler currently makes crucial and successful use of these methods. We recently tried to incorporate the connected ranges collector, but soon discovered that it bugs out in conjunction with a changing number of entities. Because of this, we expect that support will drop definitively sometime soon, and have been preparing to migrate to a different approach (using a large, fixed number of entities, of which a variable number will not participate in the actual planning). But of course, we would like to know for certain whether this is really necessary.

**Comment by triceo** [comment]
These methods are indeed and intentionally deprecated, since they aren't consistently supported across the codebase. (You mention some of the issues yourself.)

If you wish to change your problem by adding or removing entities, either do it through `ProblemChange`, or through a new instance of the solver altogether. (Under the hood, the two are very close.)

**Reply by storm-TOS** [reply]
It's not as simple as that, because finding a suitable number of entities is part of the solving process. (So we use custom moves that create and delete entities. There seems to be no way around this, including by modeling the problem in a different way etc.)

When we discussed this with Geoffrey a while back, he told us that in such cases, other users tend to use a stack of backup dummy entities during solving, which is probably our best bet as well. He said that he had never heard of users using the entity added/removed methods, but our case made him reconsider the deprecation, which is why I wanted to get back to this.

**Reply by triceo** [reply]
Yes, I have heard of this use case too. Unfortunately, this doesn't change the fact that these events are currently not reliable and making them reliable would likely require redoing many components in the solver. 

Personally, I'd like for this to work. But at the moment, there are greater priorities and marking this functionality clearly as deprecated is, IMO, preferrable to misleading users into using a feature that only partially works.

**Reply by asaveljevs** [reply]
Our VRP solvers (currently still based on OptaPlanner) also rely on the ability to add and remove entities dynamically - we discussed some use cases with @ge0ffrey at https://issues.redhat.com/browse/PLANNER-2642 in 2022.

We had some hopes for migrating to Timefold, because the documentation says "A Move can affect multiple entities or even create/delete entities." (section 3.1.1) and "A Move can only change/add/remove planning entities…" (section 3.6), so adding and removing entities was seemingly supported, but then we found this issue and it may be that upgrading to Timefold may actually not be so simple.

**Reply by triceo** [reply]
These methods are well and truly deprecated, and in Timefold Solver 2.0, they are being removed.
The solver unfortunately does not support this functionality, and I am surprised that even a subset of Solver functionality apparently works. 

All new Solver features for many years already have not supported dynamically adding or removing entities. (This includes the OptaPlanner days as well.) It may be that some pieces of this seem to work, but it is highly probable that they do in fact introduce subtle problems.

We are open to a conversation regarding bringing this feature to the solver properly. But considering that it touches literally every single piece of the core solver, this would be a very long process.
